packages/library/src/compute-limit-internal.ts

```diff
@@ -0,0 +1,187 @@
+ *     getComputeUnitEstimateForTransactionMessageFactory({ rpc });
+ *
+ * // Create your transaction message.
+ * const transactionMessage = pipe(
+ *     createTransactionMessage({ version: 'legacy' }),
+ *     /* ... *\/
+ * );
+ *
+ * // Request an estimate of the actual compute units this message will consume.
+ * const computeUnitsEstimate =
+ *     await getComputeUnitEstimateForTransactionMessage(transactionMessage);
+ *
+ * // Set the transaction message's compute unit budget.
+ * const transactionMessageWithComputeUnitLimit = prependTransactionMessageInstruction(
+ *     getSetComputeLimitInstruction({ units: computeUnitsEstimate }),
+ *     transactionMessage,
+ * );
+ * ```
+ *
+ * > [!WARNING]
+ * > The compute unit estimate is just that &ndash; an estimate. The compute unit consumption of the
+ * > actual transaction might be higher or lower than what was observed in simulation. Unless you
+ * > are confident that your particular transaction message will consume the same or fewer compute
+ * > units as was estimated, you might like to augment the estimate by either a fixed number of CUs
+ * > or a multiplier.
+ *
+ * > [!NOTE]
+ * > If you are preparing an _unsigned_ transaction, destined to be signed and submitted to the
+ * > network by a wallet, you might like to leave it up to the wallet to determine the compute unit
+ * > limit. Consider that the wallet might have a more global view of how many compute units certain
+ * > types of transactions consume, and might be able to make better estimates of an appropriate
+ * > compute unit budget.
+ */
+export async function getComputeUnitEstimateForTransactionMessage_INTERNAL_ONLY_DO_NOT_EXPORT({
+    abortSignal,
+    rpc,
+    transactionMessage,
+    ...simulateConfig
+}: ComputeUnitEstimateForTransactionMessageConfig): Promise<number> {
+    /**
+     * STEP 1: Make sure the transaction message will not fail in simulation for lack of a lifetime
+     *         - either a recent blockhash lifetime or a nonce.
+     */
+    const isDurableNonceTransactionMessage = isDurableNonceTransaction(transactionMessage);
+    let compilableTransactionMessage;
+    if (isDurableNonceTransactionMessage || isTransactionMessageWithBlockhashLifetime(transactionMessage)) {
+        compilableTransactionMessage = transactionMessage;
+    } else {
+        compilableTransactionMessage = setTransactionMessageLifetimeUsingBlockhash(
+            INVALID_BUT_SUFFICIENT_FOR_COMPILATION_BLOCKHASH,
+            transactionMessage,
+        );
+    }
+    /**
+     * STEP 2: Ensure that the message has a `SetComputeLimit` instruction. The set compute limit
+     *         instruction itself consumes compute units, so it must be included in the simulation.
+     */
+    const existingSetComputeUnitLimitInstructionIndex =
+        transactionMessage.instructions.findIndex(isSetComputeLimitInstruction);
+    const maxComputeUnitLimitInstruction = createComputeUnitLimitInstruction(4294967295 /* U32::MAX */);
```

Comment:
Speculative: this might increase the estimate because it's higher than the actual max allowed, which is 1.4M - so there might be CUs consumed normalising it to 1.4M that won't be consumed when it's set to a realistic value.

Reply:
Oh nice catch! I tried this.

|1.4M|u32::MAX|
|---|---|
|![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lE1zD7Jpg6mWNv3djNKZ/f265ff86-c992-4dc2-8a42-7eda1d25d562.png)|![image.png](https://graphite-user-uploaded-assets-prod.s3.amazonaws.com/lE1zD7Jpg6mWNv3djNKZ/17723498-2753-4600-a5f0-3ee65956c567.png)|

Same number of CUs. Looks like the `min()` gets called unconditionally, so takes the same amount?

https://github.com/anza-xyz/agave/blob/75cc70d3b6671362bf2605326b05e623b4a0e679/program-runtime/src/compute_budget_processor.rs#L128C9-L133

1. Bad: being wrong about this in the future
2. Also bad: hardcoding `MAX_COMPUTE_UNIT_LIMIT` into JS clients then not being able to take it back.

I think I'll leave it the way it is here.

Reply:
Actually the absence of an effect here probably has more to do with this.

https://github.com/anza-xyz/agave/blob/d87e23d8d91c32d5f2be2bb3557c730bee1e9434/programs/compute-budget/src/lib.rs#L3